astropy/coordinates/builtin_frames/offset.py

```diff
@@ -0,0 +1,276 @@
+
+            newname = name[:-5] if name.endswith("Frame") else name
+            newname += framecls.__name__
+
+            return super().__new__(cls, newname, bases, members)
+
+    # We need this to handle the intermediate metaclass correctly, otherwise
+    # we could just subclass OffsetFrame.
+    _OffsetFrameCls = OffsetMeta("OffsetFrame",
+                                 (OffsetFrame, framecls),
+                                 {"__doc__": OffsetFrame.__doc__})
+
+    # -----------------------------------------------------
+
+    @frame_transform_graph.transform(
+        FunctionTransform, _OffsetFrameCls, _OffsetFrameCls)
+    def offset_to_offset(from_offset_coord, to_offset_frame):
+        """Transform between two offset frames."""
+        # This transform goes through the parent frames on each side.
+        # from_frame -> from_frame.origin -> to_frame.origin -> to_frame
+        intermediate_from = from_offset_coord.transform_to(
+            from_offset_coord.origin,
+        )
+        intermediate_to = intermediate_from.transform_to(
+            to_offset_frame.origin,
+        )
+        return intermediate_to.transform_to(to_offset_frame)
+
+    # -----------------------------------------------------
+
+    @frame_transform_graph.transform(AffineTransform, framecls, _OffsetFrameCls)
+    def reference_to_offset(reference_coord, offset_frame):
+        """Convert a reference coordinate to an offset frame."""
+        _check_coord_repr_diff_types(reference_coord)
+        return get_matrix_vectors(offset_frame)
+
+    # -----------------------------------------------------
+
+    @frame_transform_graph.transform(AffineTransform, _OffsetFrameCls, framecls)
+    def offset_to_reference(offset_coord, _):
+        """Convert an offset frame coordinate to the reference frame."""
+        _check_coord_repr_diff_types(offset_coord)
+        return get_matrix_vectors(offset_coord, inverse=True)
+
+    # -----------------------------------------------------
+
+    def get_matrix_vectors(offset_frame, inverse=False):
+        """Get matrix vectors of offset.
+
+        Use the ``inverse`` argument to get the inverse transformation, matrix
+        and offsets to go from OffsetFrame to from_frame.
+
+        .. todo::  allow for rotations about the axes
+
+        """
+        of = offset_frame  # shorthand
+
+        # # construct transformation matrix and use it
+        R = np.eye(3)  # no rotation
+
+        # compute total matrices
+        A = R
+
+        if of.origin.data.differentials:
+            translation = of.origin.represent_as(r.CartesianRepresentation,
+                                                 s=r.CartesianDifferential)
+        else:
+            translation = of.origin.represent_as(r.CartesianRepresentation)
+            translation = translation.with_differentials(zero_v)
+
+        offset = -translation.without_differentials()
```

Comment:
There are / have been some PRs to enable doing this **with** differentials attached.